### PR TITLE
Tracing: Support remote, rate-limited, and probabilistic sampling in tracing.opentelemetry config section

### LINKED ISSUE #43947: Extend configurability of OpenTelemetry
After OpenTelemetry was added as a configurable option for tracing we can extend the features and add configuration for them

- [ ] Add remote sampling
- [ ] Add rate limiting sampling 
- [ ] Add tempo as an alternative to jaeger

### PR DESCRIPTION
**Pull request by hairyhenderson**
**What is this feature?**

Supports sampling configuration in the `tracing.opentelemetry` config section. All sampling strategies that were supported with Jaeger are supported now with all OTel providers (Jaeger & OTLP), especially the remote sampling server.

**Why do we need this feature?**

The `tracing.jaeger` section is deprecated, but there's no way to configure sampling otherwise.

**Who is this feature for?**

Operators

**Which issue(s) does this PR fix?**:

Fixes #43947

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 1]
```diff
@@ -1797,6 +1797,8 @@ Refer to https://www.jaegertracing.io/docs/1.16/sampling/#client-sampling-config
 
 Can be set with the environment variable `JAEGER_SAMPLER_TYPE`.
 
+_Setting `sampler_type` in the `tracing.opentelemetry` section will override this setting._
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1800]:
```suggestion
_To override this setting, enter `sampler_type` in the `tracing.opentelemetry` section._
```

Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1800]: @hairyhenderson , wanted to check with you about the above suggestion before I replicate that wording convention to the docs below.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
 
 Can be set with the environment variable `OTEL_RESOURCE_ATTRIBUTES` (use `=` instead of `:` with the environment variable).
 
+### sampler_type
+
+Default value is `const`.
+
+Specifies the type of sampler: `const`, `probabilistic`, `ratelimiting`, or `remote`.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1855]:
```suggestion
This option specifies the type of sampler: `const`, `probabilistic`, `ratelimiting`, or `remote`.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 3]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+
+### sampler_param
+
+Default value is `1`.
+
+Sampler configuration parameter. Depending on the value of `sampler_type`, it can be `0`, `1`, or a decimal value in between.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1861]:
```suggestion
Depending on the value of `sampler_type`, the sampler configuration parameter can be `0`, `1`, or any decimal value between `0` and `1`.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 4]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+
+- For `const` sampler, `0` or `1` to always `false`/`true` respectively
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1863]:
```suggestion
- For `const` sampler, `0` or `1` means `false`/`true` respectively
```

Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1863]: Does that suggestion  ^ make sense?

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+- For `probabilistic` sampler, a probability between `0.0` and `1.0`
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1864]:
```suggestion
- For the `probabilistic` sampler, you can use a decimal value between `0.0` and `1.0`
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 6]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+- For `rateLimiting` sampler, the number of spans per second
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1865]:
```suggestion
- For the `rateLimiting` sampler, enter the number of spans per second
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 7]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+- For `remote` sampler, param is the same as for `probabilistic`
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1866]:
```suggestion
- For the `remote` sampler, enter the same parameter value as you entered for `probabilistic`.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 8]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+  and indicates the initial sampling rate before the actual one
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1867]:
```suggestion
 the initial sampling rate before a sample is received from the sampling server.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 9]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+  is received from the sampling server
+
+### sampling_server_url
+
+When `sampler_type` is `remote`, this specifies the URL of the sampling server. This can be used by all tracing providers. 
+
+The sampling server should implement the Jaeger remote sampling API, which can be exposed by a jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or the Grafana Agent.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1874]: does the user do this implementation? I'm wondering if this should be something like "In the sampling server, implement the Jaeger remote sampling API...".

**Review by chri2547 — CHANGES_REQUESTED** [review]
@hairyhenderson , added some suggestions and a couple of questions.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 4 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1863]: yep - the "always" wording was perhaps trying to be too succinct. What was intended is that the sampling decision would always be `false` or `true` depending on the value.

There's probably a better way to word it. How about:

```suggestion
- For `const` sampler, use `0` to never sample or `1` to always sample
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: No, the user just needs to choose a server that implements that API - the 4 listed all implement the Jaeger remote sampling API.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 7 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1866]: This doesn't make sense - the new wording implies the value should match what was entered for `probabilistic`, except it's an either-or choice - you can't choose both the `remote` _and_ `probabilistic` sampler.

How about:

```suggestion
- For the `remote` sampler, use a decimal value between `0.0` and `1.0` to specify
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 8 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1867]: This no longer makes sense as samples are not received from the sampling server...

Also, a multi-line suggestion would be _much_ easier to reason through 😅 I'm going to start a new one with a counter-proposal.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 10]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+Default value is `1`.
+
+Depending on the value of `sampler_type`, the sampler configuration parameter can be `0`, `1`, or any decimal value between `0` and `1`.
+
+- For `const` sampler, `0` or `1` to always `false`/`true` respectively
+- For the `probabilistic` sampler, you can use a decimal value between `0.0` and `1.0`
+- For the `rateLimiting` sampler, enter the number of spans per second
+- For `remote` sampler, param is the same as for `probabilistic`
+  and indicates the initial sampling rate before the actual one
+  is received from the sampling server
```
Comment (hairyhenderson) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1868]: @chri2547 how about this:

```suggestion
- For the `remote` sampler, use a decimal value between `0.0` and `1.0`
  to specify the initial sampling rate used before the first update
  is received from the sampling server
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 8 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1867]: https://github.com/grafana/grafana/pull/73587/files#r1320166327

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 7 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1866]: See https://github.com/grafana/grafana/pull/73587/files#r1320166327

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: 👍 , all good on this one then.

Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: How about:

```suggestion
Use a sampling server that implement the Jaeger remote sampling API, such as jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or Grafana Agent.
```

(maybe `Grafana Agent` should link to https://grafana.com/oss/agent/?)

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 7 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1866]: Ah, that works for me. Thank you.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 11]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+
+Default value is `1`.
+
+Depending on the value of `sampler_type`, the sampler configuration parameter can be `0`, `1`, or any decimal value between `0` and `1`.
+
+- For `const` sampler, `0` or `1` to always `false`/`true` respectively
+- For the `probabilistic` sampler, you can use a decimal value between `0.0` and `1.0`
+- For the `rateLimiting` sampler, enter the number of spans per second
+- For `remote` sampler, param is the same as for `probabilistic`
+  and indicates the initial sampling rate before the actual one
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1867]:
```suggestion
  and indicate the initial sampling rate before the actual one
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 4 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1863]: 👍 , that works for me, thank you.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 2 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1855]: hmm... is it "option" or "setting"? I see both used in a few places in this doc, but most entries are just implicit (that's why I used the original form)

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 10 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1868]: Perfect, thank you.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: @chri2547 I think I was busy writing this comment and didn't see your original reply - what do you think of the alternate phrasing?

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 2 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1855]: Good point, it's probably "setting" but we can leave this as is so as not to introduce an inconsistency in the docs. I'm fine with "Specifies the type...".

Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1855]: ok - TBH there's already inconsistency - maybe that's worth cleaning up _later_ 😉

I'll leave this one alone for now.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: I like that, tho would use "to" instead of "that" - so "Use a sampling server to implement..."

**Review by chri2547 — APPROVED** [review]
LGTM, thanks for the contribution!

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: oops - it should've been `that implements`

"Use a sampling server to implement ..." doesn't really scan well to me - the user's really just running some server and putting its URL in this setting. The important thing is the server needs to support the correct API. Actually, maybe "supports" is a clearer term. Or maybe I'm bikeshedding myself at this point 😂

Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: How about:

```suggestion
Use a sampling server that supports the Jaeger remote sampling API, such as jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or [Grafana Agent](https://grafana.com/oss/agent/).
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 2 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1855]: 👍

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 9 above)
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1874]: 👍

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 12]
```diff
@@ -1842,6 +1848,31 @@ Comma-separated list of attributes to include in all new spans, such as `key1:va
...
+
+### sampling_server_url
+
+When `sampler_type` is `remote`, this specifies the URL of the sampling server. This can be used by all tracing providers. 
+
+Use a sampling server that supports the Jaeger remote sampling API, such as jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or [Grafana Agent](https://grafana.com/oss/agent/).
+
 <hr>
 
 ## [tracing.opentelemetry.jaeger]
```
[the remark this answers is not in the document]
Comment (hairyhenderson) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1878]: huh, TIL 😂

**Review by marefr — APPROVED** [review]
pkg/plugins/localfiles.go change is fine. Other changes LGTM as well, but would like @grafana/backend-platform to have a look as well since they're codeowners

File: pkg/plugins/localfiles.go [hunk 13]
```diff
@@ -33,7 +33,7 @@ func NewLocalFS(basePath string) LocalFS {
 // fileIsAllowed takes an absolute path to a file and an os.FileInfo for that file, and it checks if access to that
 // file is allowed or not. Access to a file is allowed if the file is in the FS's Base() directory, and if it's a
 // symbolic link it should not end up outside the plugin's directory.
-func (f LocalFS) fileIsAllowed(basePath string, absolutePath string, info os.FileInfo) (bool, error) {
+func (f LocalFS) fileIsAllowed(basePath string, absolutePath string, info fs.FileInfo) (bool, error) {
```
Comment (xnyo) [inline pkg/plugins/localfiles.go:36]: No big deal since

```go
package os 

// ...

type FileInfo = fs.FileInfo
```

But why was this changed?

File: pkg/plugins/localfiles.go [hunk 14]
```diff
@@ -33,7 +33,7 @@ func NewLocalFS(basePath string) LocalFS {
 // fileIsAllowed takes an absolute path to a file and an os.FileInfo for that file, and it checks if access to that
```
Comment (xnyo) [inline pkg/plugins/localfiles.go:33]:
```suggestion
// fileIsAllowed takes an absolute path to a file and an fs.FileInfo for that file, and it checks if access to that
```

**Review by xnyo — APPROVED** [review]
Left some nits but LGTM!

This will probably also require changes in the plugin-sdk so the sampler settings are propagated properly:

https://github.com/grafana/grafana-plugin-sdk-go/blob/05a3321740553735a6181e1491a1a02547973a7f/internal/tracerprovider/tracerprovider.go#L72

But I can open an issue and do that separately :)

File: pkg/plugins/localfiles.go (hunk 13 above)
Comment (hairyhenderson) [reply pkg/plugins/localfiles.go:36]: I don't recall my exact thinking at the time, but `fs.FileInfo` is the "more correct" one. I may have been thinking about refactoring this to use `io/fs` rather than `os` wherever I could for easier injection of in-memory filesystems at test-time.

However, given that this is the only change in this file, I'm going to back it out - it's not relevant to this PR so the change just confuses history.

File: pkg/plugins/localfiles.go (hunk 14 above)
Comment (hairyhenderson) [reply pkg/plugins/localfiles.go:33]: see https://github.com/grafana/grafana/pull/73587#discussion_r1321754124 - I'm going to back out the change so I won't change this either.

**Comment by hairyhenderson** [comment]
> This will probably also require changes in the plugin-sdk so the sampler settings are propagated properly:
> 
> https://github.com/grafana/grafana-plugin-sdk-go/blob/05a3321740553735a6181e1491a1a02547973a7f/internal/tracerprovider/tracerprovider.go#L72
> 
> But I can open an issue and do that separately :)

@xnyo ah - thanks for that.
